pil_analyzer/src/evaluator.rs

```diff
@@ -1,104 +1,394 @@
+    parsed::{display::quote, FunctionCall, MatchArm, MatchPattern},
 };
+use itertools::Itertools;
 use number::FieldElement;
 
-/// Evaluates an expression to a single value.
+/// Evaluates an expression to a single number.
 pub fn evaluate_expression<T: FieldElement>(
-    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition<T>>)>,
     expression: &Expression<T>,
-) -> Result<T, String> {
-    Evaluator {
-        definitions,
-        function_cache: &Default::default(),
-        variables: &[],
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition<T>>)>,
+) -> Result<T, EvalError> {
+    internal::evaluate_to_number::<T, NoCustom>(expression, &[], &Definitions(definitions))
+}
+
+pub fn evaluate<'a, T: FieldElement, C: Custom>(
+    expr: &'a Expression<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
+) -> Result<Value<'a, T, C>, EvalError> {
+    internal::evaluate(expr, &[], symbols)
+}
+
+pub fn evaluate_function_call<'a, T: FieldElement, C: Custom>(
+    expr: &'a Expression<T>,
+    arguments: Vec<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
+) -> Result<Value<'a, T, C>, EvalError> {
+    internal::evaluate(
+        expr,
+        &arguments
+            .into_iter()
+            .map(|x| Rc::new(Value::Number(x)))
+            .collect::<Vec<_>>(),
+        symbols,
+    )
+}
+
+pub fn evaluate_to_number<'a, T: FieldElement, C: Custom>(
+    expr: &'a Expression<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
+) -> Result<T, EvalError> {
+    internal::evaluate_to_number(expr, &[], symbols)
+}
+
+pub fn evaluate_function_call_to_number<'a, T: FieldElement, C: Custom>(
+    expr: &'a Expression<T>,
+    arguments: Vec<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
+) -> Result<T, EvalError> {
+    internal::evaluate_to_number(
+        expr,
+        &arguments
+            .into_iter()
+            .map(|x| Rc::new(Value::Number(x)))
+            .collect::<Vec<_>>(),
+        symbols,
+    )
+}
+
+/// Evaluation errors.
+/// TODO Most of these errors should be converted to panics as soon as we have a proper type checker.
+#[derive(Debug)]
+pub enum EvalError {
+    /// Type error, for example non-number used as array index.
+    TypeError(String),
+    /// Fundamentally unsupported operation (regardless of type), e.g. access to public variables.
+    Unsupported(String),
+    /// Array index access out of bounds.
+    OutOfBounds(String),
+    /// Unable to match pattern. TODO As soon as we have "Option", patterns should be exhaustive
+    /// This error occurs quite often and thus should not require allocation.
+    NoMatch(),
+    /// Reference to an undefined symbol
+    SymbolNotFound(String),
+    /// Data not (yet) available
+    DataNotAvailable,
+}
+
+#[derive(Clone, PartialEq)]
+pub enum Value<'a, T, C> {
+    Number(T),
+    String(String),
+    Tuple(Vec<Self>),
+    Array(Vec<Self>),
+    Closure(Closure<'a, T, C>),
+    Custom(C),
```

Comment:
No, it can be anything. The idea is that this file implement the basic evaluator on a minimalistic functional language. The `Custom` and `SymbolLookup` mechanism allows the language to be extended. The extension is already different in the query_processor and in the constant_evaluator. Another extension will be added in https://github.com/powdr-labs/powdr/pull/773 where the condenser uses this enum alternative to either store an AlgebraicExpression or a Constraint.

So essentially all the other types can be seen as "built-in" core language and this one is the "user-defined" part.